File: src/lib/events.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { get } from "svelte/store";
import {
  backingRms,
  duration,
  elapsed,
  micRms,
  pausedAtElapsed,
  pausedResumeMode,
  transportState,
  type TransportState,
} from "../stores/transport";
import {
  backingPitchAnalyzing,
  backingPitchQuality,
  backingPitchTrack,
  currentPitch,
  freeMode,
  freeModeReason,
  liveVocalSamples,
  type BackingPitchQuality,
  type PitchSample,
  type PitchTrack,
} from "../stores/pitch";
import { showToast } from "../stores/toast";

interface BackingPitchAnalyzingPayload {
  duration: number;
}

interface BackingPitchNotDetectedPayload {
  voiced_ratio: number;
  mean_confidence: number;
  elapsed_secs: number;
  reason: string;
}

let unlisteners: UnlistenFn[] = [];

export async function setupEventListeners(): Promise<void> {
  await teardownEventListeners();

  unlisteners.push(
    await listen<{ elapsed: number; duration: number }>("audio:progress", (e) => {
      elapsed.set(e.payload.elapsed);
      duration.set(e.payload.duration);
    }),
  );

  unlisteners.push(
    await listen<{ backing_rms: number; mic_rms: number }>("audio:rms", (e) => {
      backingRms.set(e.payload.backing_rms);
      micRms.set(e.payload.mic_rms);
    }),
  );

  unlisteners.push(
    await listen<{ state: string }>("audio:state_changed", (e) => {
      const state = e.payload.state;
      if (state === "idle") return;
      transportState.set(state as TransportState);
    }),
  );

  unlisteners.push(
    await listen<PitchSample | null>("audio:pitch", (e) => {
      currentPitch.set(e.payload);
      if (!e.payload) return;

      const timestamp = get(elapsed);
      liveVocalSamples.update((samples) => {
        if (samples.length >= 600) {
          samples.shift();
        }
        samples.push({
          timestamp,
          freq: e.payload!.freq,
          confidence: e.payload!.confidence,
          note: e.payload!.note,
          octave: e.payload!.octave,
          cent: e.payload!.cent,
        });
        return samples;
      });
    }),
  );

  unlisteners.push(
    await listen<BackingPitchAnalyzingPayload>("backing_pitch:analyzing", (e) => {
      backingPitchAnalyzing.set({ duration: e.payload.duration });
      backingPitchTrack.set(null);
      freeMode.set(false);
      freeModeReason.set(null);
    }),
  );

  unlisteners.push(
    await listen<BackingPitchQuality>("backing_pitch:ready", async (e) => {
      backingPitchAnalyzing.set(null);
      try {
        const track = await invoke<PitchTrack | null>("get_backing_pitch_track");
        if (track) {
          backingPitchTrack.set(track);
          backingPitchQuality.set(e.payload);
          freeMode.set(false);
          freeModeReason.set(null);
        }
      } catch (err) {
        console.warn("[backing_pitch] failed to load pitch track:", err);
        backingPitchTrack.set(null);
        freeMode.set(true);
        freeModeReason.set({
          kind: "i18n",
          key: "pitchTimeline.banner.freeMode.loadFailed",
        });
      }
    }),
  );

  unlisteners.push(
    await listen<BackingPitchNotDetectedPayload>("backing_pitch:not_detected", (e) => {
      backingPitchAnalyzing.set(null);
      backingPitchTrack.set(null);
      backingPitchQuality.set({
        total_frames: 0,
        voiced_frames: 0,
        voiced_ratio: e.payload.voiced_ratio,
        mean_confidence: e.payload.mean_confidence,
        elapsed_secs: e.payload.elapsed_secs,
      });
      freeMode.set(true);
      freeModeReason.set({ kind: "text", text: e.payload.reason });
    }),
  );

  unlisteners.push(
    await listen("audio:finished", () => {
      if (get(transportState) === "paused") return;
      transportState.set("idle");
      pausedResumeMode.set(null);
      pausedAtElapsed.set(null);
    }),
  );

  unlisteners.push(
    await listen<{ message: string }>("audio:error", (e) => {
      console.error("[audio:error]", e.payload.message);
      showToast(e.payload.message, "error", 5000);
      transportState.set("idle");
      pausedResumeMode.set(null);
      pausedAtElapsed.set(null);
    }),
  );
}

export async function teardownEventListeners(): Promise<void> {
  for (const fn of unlisteners) {
    fn();
  }
  unlisteners = [];
}

File: src-tauri/src/events.rs
use serde::Serialize;
use tauri::{AppHandle, Emitter};

pub const AUDIO_PROGRESS: &str = "audio:progress";
pub const AUDIO_RMS: &str = "audio:rms";
pub const AUDIO_PITCH: &str = "audio:pitch";
pub const AUDIO_STATE_CHANGED: &str = "audio:state_changed";
pub const AUDIO_FINISHED: &str = "audio:finished";
pub const AUDIO_ERROR: &str = "audio:error";
pub const BACKING_PITCH_ANALYZING: &str = "backing_pitch:analyzing";
pub const BACKING_PITCH_READY: &str = "backing_pitch:ready";
pub const BACKING_PITCH_NOT_DETECTED: &str = "backing_pitch:not_detected";

#[derive(Clone, Serialize)]
pub struct ProgressPayload {
    pub elapsed: f64,
    pub duration: f64,
}

#[derive(Clone, Serialize)]
pub struct RmsPayload {
    pub backing_rms: f32,
    pub mic_rms: f32,
}

#[derive(Clone, Serialize)]
pub struct PitchPayload {
    pub freq: f64,
    pub note: String,
    pub octave: i32,
    pub cent: f64,
    pub confidence: f64,
}

#[derive(Clone, Serialize)]
pub struct StatePayload {
    pub state: String,
}

#[derive(Clone, Serialize)]
pub struct ErrorPayload {
    pub message: String,
}

#[derive(Clone, Serialize)]
pub struct BackingPitchAnalyzingPayload {
    pub duration: f64,
}

#[derive(Clone, Serialize)]
pub struct BackingPitchQualityPayload {
    pub total_frames: usize,
    pub voiced_frames: usize,
    pub voiced_ratio: f64,
    pub mean_confidence: f64,
    pub elapsed_secs: f64,
}

#[derive(Clone, Serialize)]
pub struct BackingPitchNotDetectedPayload {
    pub voiced_ratio: f64,
    pub mean_confidence: f64,
    pub elapsed_secs: f64,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct CalibrationCompletePayload {
    pub mode: String,
    pub latency_ms: u64,
    pub confidence: String,
    pub rounds_used: u8,
    pub valid_beats: u8,
    pub measurement_beats: u8,
    pub std_dev_ms: f64,
    pub round_spread_ms: f64,
    pub applied_recommended: bool,
    pub diagnostic: String,
}

pub fn emit_progress(app: &AppHandle, elapsed: f64, duration: f64) {
    let _ = app.emit(AUDIO_PROGRESS, ProgressPayload { elapsed, duration });
}

pub fn emit_rms(app: &AppHandle, backing_rms: f32, mic_rms: f32) {
    let _ = app.emit(
        AUDIO_RMS,
        RmsPayload {
            backing_rms,
            mic_rms,
        },
    );
}

pub fn emit_state(app: &AppHandle, state: &str) {
    let _ = app.emit(
        AUDIO_STATE_CHANGED,
        StatePayload {
            state: state.to_string(),
        },
    );
}

pub fn emit_pitch(app: &AppHandle, payload: PitchPayload) {
    let _ = app.emit(AUDIO_PITCH, payload);
}

pub fn emit_pitch_none(app: &AppHandle) {
    let _ = app.emit(AUDIO_PITCH, serde_json::Value::Null);
}

pub fn emit_finished(app: &AppHandle) {
    let _ = app.emit(AUDIO_FINISHED, ());
}

pub fn emit_error(app: &AppHandle, message: &str) {
    let _ = app.emit(
        AUDIO_ERROR,
        ErrorPayload {
            message: message.to_string(),
        },
    );
}

pub fn emit_backing_pitch_analyzing(app: &AppHandle, payload: BackingPitchAnalyzingPayload) {
    let _ = app.emit(BACKING_PITCH_ANALYZING, payload);
}

pub fn emit_backing_pitch_ready(app: &AppHandle, payload: BackingPitchQualityPayload) {
    let _ = app.emit(BACKING_PITCH_READY, payload);
}

pub fn emit_backing_pitch_not_detected(app: &AppHandle, payload: BackingPitchNotDetectedPayload) {
    let _ = app.emit(BACKING_PITCH_NOT_DETECTED, payload);
}
